Nodejs/Product/Nodejs/SharedProject/CommonProjectNode.DiskMerger.cs

```diff
@@ -12,128 +15,226 @@ internal abstract partial class CommonProjectNode
         /// Performs merging of the file system state with the current project hierarchy, bringing them
         /// back into sync.
         /// 
-        /// The class can be created, and ContinueMerge should be called until it returns false, at which
-        /// point the file system has been merged.  
+        /// The worker thread runs continuously to sync the files on disk with the DiskNodes collection of 
+        /// the project.
         /// 
-        /// You can wait between calls to ContinueMerge to enable not blocking the UI.
+        /// Callers can call the Wait method to blocking wait on the completion of the merge, i.e. till the collection 
+        /// is empty.
         /// 
-        /// If there were changes which came in while the DiskMerger was processing then those changes will still need
-        /// to be processed after the DiskMerger completes.
+        /// Checks for IsDisposed should be done before calling any of the methods of the BlockingCollection, which could
+        /// throw ObjectDisposed exceptions otherwise.
         /// </summary>
-        private sealed class DiskMerger
+        private sealed class DiskMerger : IDisposable
         {
-            private readonly Stack<DirState> remainingDirs = new Stack<DirState>();
+            private readonly BlockingCollection<DirState> dirs = new BlockingCollection<DirState>(new ConcurrentStack<DirState>());
+
             private readonly CommonProjectNode project;
 
-            public DiskMerger(CommonProjectNode project, HierarchyNode parent, string dir)
+            private readonly Thread worker;
+
+            private readonly ManualResetEventSlim processingDirsEvent = new ManualResetEventSlim(initialState: true);
```

Comment:
Why use the *Slim rather than just ManualResetEvent? Slim uses spinning and is for when wait times are expected to be very short, which is probably unlikely if processing all directories.